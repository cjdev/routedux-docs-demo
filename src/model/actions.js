
export const CHANGE_ID = "CHANGE_ID";
export const CONTENT_RECEIVED = "CONTENT_RECEIVED";

const createContentReceived = (id, contents) => {
  return {type: CONTENT_RECEIVED, id, contents};
};

const changeId = (id) => {
  return { type: CHANGE_ID, id };
};

export const actions = {
  changeId,
  createContentReceived
};