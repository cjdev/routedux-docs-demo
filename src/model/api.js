
export const loadArticle = (function () {
  async function loadContents(url) {
    const response = await fetch(url);

    if (response.status !== 200) {
      throw new Error({response: response.status});
    }
    return response.text();
  }

  const contents = {};

  // this is basically fetch with caching.
  return function(url) {
    if (contents[url]) {
      return contents[url];
    }
    return contents[url] = loadContents(url);
  }
})();