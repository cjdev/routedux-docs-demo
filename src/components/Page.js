import React from 'react';
import Layout from './Layout';
import {connect} from 'react-redux';
import {actions, getContentsById} from '../model';

function Page({menu, title, contents}) {
  return (
    <Layout nav={menu}
            content={<div dangerouslySetInnerHTML={{__html: contents}}/>}
            headerContent={title}/>
  );
}

const mapStateToProps = ({currentTitle, id, ...state}, own) => {

  let contents;
  const title = own.title ? own.title : currentTitle;

  if(own.contents) {
    contents = own.contents;
  }
  if(!contents) {
    const foundContents = getContentsById(id, state);
    contents = foundContents ? foundContents : 'Loading...';
  }

  return {menu: own.menu, title, contents};
};

const mapActionsToProps = {
  needContent: actions.changeId
};

export default connect(mapStateToProps, mapActionsToProps)(Page);

export {Page, mapStateToProps};
