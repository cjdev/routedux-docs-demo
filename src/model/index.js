import {
  createStore,
  applyMiddleware,
  compose
} from 'redux';

import createSagaMiddleware from 'redux-saga';
import {call, put, takeEvery} from 'redux-saga/effects'

import R from 'ramda';

import pages from '../data/pages.json';

import {
  CHANGE_ID,
  CONTENT_RECEIVED,
  actions as allActions
} from './actions';

import {loadArticle} from './api';

import {installBrowserRouter} from 'routedux';

const createContentReceived = allActions.createContentReceived;

const createDefaultState = () => {
  return {
    id: "home",
    currentTitle: false,
    pages: pages,
    contentMap: {}
  };
}

const defaultState = createDefaultState();

// pages = [{ id, contentUrl, title}]

const lookupPageFromDirectory = (state, id) => {
  const matcher = R.propEq('id', id);
  return R.find(matcher)(state.pages);
};

const lookupTitle = (state, id) => {
  const pageData = lookupPageFromDirectory(state, id);
  return (pageData && pageData.title) || null;
};

const doUpdateContent = (state, event) => {
  const id = event.id;
  const contents = event.contents;
  let newState = Object.assign({}, state);

  newState.id = id;
  newState.contentMap = Object.assign({}, state.contentMap, {[id]: contents});
  newState.currentTitle = lookupTitle(state, id);

  return newState;
};

const reduce = (state = defaultState, event) => {
  let newState;
  switch (event.type) {
    case CONTENT_RECEIVED:
      newState = doUpdateContent(state, event);
      break;
    default:
      newState = state;
      break;
  }

  return newState;

};

const {middleware, enhancer, init} = installBrowserRouter([
  ['/pages/:id', 'CHANGE_ID', {}],
  ['/', 'CHANGE_ID', {id: 'home'}]
]);

// create the saga middleware
const sagaMiddleware = createSagaMiddleware();

// redux debugging.
const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
// const composeEnhancers = compose;

// create store
const store = createStore(
  reduce,
  composeEnhancers(
    enhancer,
    applyMiddleware(
      middleware,
      sagaMiddleware
    )
  ));

const fetchArticle = function*({id}) {
  console.log('SAGA???');
  try {
    const content = yield call(loadArticle, `/pages/${id}.html`);
    yield put(createContentReceived(id, content));
  } catch (err) {
    console.error("Couldn't get article:", err)
  }
};

const sagas = {fetchArticle};

const saga = function *() {
  yield takeEvery(CHANGE_ID, fetchArticle);
};

sagaMiddleware.run(saga);
// this needs to be after sagaMiddleware.run so that sagas can handle our route-dispatched actions.
init();

const actions = {
  changeId: allActions.changeId
};

init();
export default store;

// store dependent helper function - thus defined after default export
const getContentsById = (id, state = null) => {
  if (!state) {
    state = store.getState();
  }
  return state.contentMap[id] || null;
};

export {
  getContentsById,
  reduce,
  actions,
  createDefaultState,
  sagas
};
