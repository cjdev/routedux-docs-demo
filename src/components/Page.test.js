import React from 'react';
import ezJson from 'enzyme-to-json';
import {shallow} from 'enzyme';
import {Page, RouteAwarePageBuilder, mapStateToProps} from './Page';
import {actions} from '../model';

it("should render menu and contents", () => {
  //given
  ;
  //when
  const wrapper = shallow(<Page title="Title!" contents="<div>Raw Html</div>" menu="Menu contents" />);

  //then
  expect(ezJson(wrapper)).toMatchSnapshot();

});


it("RouteAwarePage should load content using id from route and fire route transition action", () => {
  //given
  let idUsed;
  const stubGetContents = (id) => {
    idUsed = id;
    return "Some content";
  };
  const stubStore = {
    actions: [],
    dispatch: function(action) {this.actions.push(action) }
  };
  const RouteAwarePage = RouteAwarePageBuilder(stubGetContents, stubStore);
  const match = {params: {id: 'routeId'}};

  //when
  const wrapper = shallow(<RouteAwarePage match={match} menu="Menu" title="Title" />);

  //then
  expect(stubStore.actions[0]).toEqual(actions.changeId('routeId'));
  expect(idUsed).toEqual('routeId');
  expect(ezJson(wrapper)).toMatchSnapshot();

});
it("RouteAwarePage should load content using passed id instead when it is present", () => {
//given
  let idUsed;
  const stubGetContents = (id) => {
    idUsed = id;
    return "Some content";
  };
  const stubStore = {
    actions: [],
    dispatch: function(action) {this.actions.push(action) }
  };
  const RouteAwarePage = RouteAwarePageBuilder(stubGetContents, stubStore);
  const match = {params: {id: 'routeId'}};

  //when
  const wrapper = shallow(<RouteAwarePage id="ownId" match={match} menu="Menu" title="Title" />);

  //then
  expect(stubStore.actions[0]).toEqual(actions.changeId('ownId'));
  expect(idUsed).toEqual('ownId');
  expect(ezJson(wrapper)).toMatchSnapshot();
});

it("RouteAwarePage uses passed contents by default when content cannot be loaded", () => {
  let idUsed;
  const stubGetContents = (id) => {
    idUsed = id;
    return null;
  };
  const stubStore = {
    actions: [],
    dispatch: function(action) {this.actions.push(action) }
  };
  const RouteAwarePage = RouteAwarePageBuilder(stubGetContents, stubStore);
  const match = {params: {id: 'routeId'}};

  //when
  const wrapper = shallow(<RouteAwarePage contents="Hello" match={match} menu="Menu" title="Title" />);

  //then
  expect(ezJson(wrapper)).toMatchSnapshot();
});

it("mapStateToProps uses its own title and contents preferentially", () => {
  // given
  const state = {currentTitle: 'currentTitle', id: 'foo', contentMap: {'foo' : 'Foo Contents'}};
  const ownProps = {title: 'Own Title', contents: 'Own Contents', menu: 'Menu'};
  const expectedProps = {title: 'Own Title', contents: 'Own Contents', menu: 'Menu'};

  // when
  const returnedProps = mapStateToProps(state, ownProps);

  //then
  expect(returnedProps).toEqual(expectedProps);

});