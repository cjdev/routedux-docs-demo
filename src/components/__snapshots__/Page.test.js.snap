// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RouteAwarePage should load content using id from route and fire route transition action 1`] = `
<Connect(Page)
  contents="Some content"
  menu="Menu"
  title="Title"
/>
`;

exports[`RouteAwarePage should load content using passed id instead when it is present 1`] = `
<Connect(Page)
  contents="Some content"
  menu="Menu"
  title="Title"
/>
`;

exports[`RouteAwarePage uses passed contents by default when content cannot be loaded 1`] = `
<Connect(Page)
  contents="Hello"
  menu="Menu"
  title="Title"
/>
`;

exports[`should render menu and contents 1`] = `
<Layout
  content={
    <div
      dangerouslySetInnerHTML={
        Object {
          "__html": "<div>Raw Html</div>",
        }
      }
    />
  }
  headerContent="Title!"
  nav="Menu contents"
/>
`;
